[KEP 5080]Ordered Namespace Deletion

cici37 opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Add the KEP for ordered namespace deletion.

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/5080

<!-- other comments or additional information -->
- Other comments:

cici37 commented:
/cc @thockin @jpbetz @deads2k @aojea

thockin commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+- Monitor the deletion process for each `NamespaceDeletionOrder`. If the process hangs beyond a predefined timeout (e.g., 5 minutes), 
+  it will detect the stall and trigger the deletion attempt for the next `NamespaceDeletionOrder` group.
+
+- If the resources from the next `NamespaceDeletionOrder` group are deleted successfully, retry with the ordered namespace deletion mechanism. If the process hangs again, wait for a predefined timeout and retry.
```
Is there anything to "retry"?

Assume your order says all A's before B's before C's.  You delete all A objects, but at least one is still hanging around.  You hit the timeout and move on to B.  You delete all B objects.  Now you are waiting for all A || B objects.  If you hit the timeout, you move onto C.

Right?

cici37 replied:
Rephrased it.

thockin commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+
+- If the resources from the next `NamespaceDeletionOrder` group are deleted successfully, retry with the ordered namespace deletion mechanism. If the process hangs again, wait for a predefined timeout and retry.
+
+- If the retry fails over a certain number or the process hangs for too long(an overall timeout), it falls into the fallback mechanism.
```
Do we really need an overall timeout?  If we have 4 bands and the very first one gets stuck, you have at worst 20 minutes before you are in the exact same situation as today.

thockin reviewed (commented):
Thanks for picking this one up.

I am LGTM but for some smaller details.

brakthehack commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+- Attempt to honor the DeletionOrderPriority for resource deletion.
+
+- Monitor the deletion process for each `NamespaceDeletionOrder`. If the process hangs beyond a predefined timeout (e.g., 5 minutes), 
+  it will detect the stall and trigger the deletion attempt for the next `NamespaceDeletionOrder` group.
```
Does this algorithm violate Goal #2? For example, if someone crafted a vulnerability that somehow prevents the pod from being deleted, could they force the system to delete the policies ahead of the pods thereby re-opening the original problem statement?

I you intend this to be used for security purposes, I suspect that enforcing the invariant in all cases would prevent it from abuse at the cost of user convenience.

thockin replied:
As @danwinship also said something similar, let's discuss.

If we treat this as an inviolable ordering, we do have a risk of introducing a dep cycle which requires human action to break.  In particular "it worked fine in 1.32 and broke in 1.33".  We can argue that this is "safer" but that's small consolation to the users whose stuff broke.

If we treat this as a strong but not inviolable ordering, we avoid cycles (at the cost of time) but some attacks remain possible.  It's no worse than today, but we are trying to close the hole.

So which is worse?

Another course could be to do timeouts today, publish metrics on how often they fail, increase those timeouts over a few releases and then move it to inviolable.  E.g. .1.33 is 5min; 1.34 is 10min; 1.35 is 30min; 1.36 is infinite.  Anyone for whom this is a real issue will have time to scream and/or fix the broken pattern.

Another course would be to make it admin-configured, but I would be against that.  Mentioning it for completeness.

BenTheElder replied:
> If we treat this as an inviolable ordering, we do have a risk of introducing a dep cycle which requires human action to break. In particular "it worked fine in 1.32 and broke in 1.33". We can argue that this is "safer" but that's small consolation to the users whose stuff broke.

What usage, concretely, do we think we're breaking? I think it's easier to weigh this with realistic examples.

> If we treat this as a strong but not inviolable ordering, we avoid cycles (at the cost of time) but some attacks remain possible. It's no worse than today, but we are trying to close the hole.

Well, playing devil's advocate ...  It's worse in that people may have a false sense of security that "we fixed the problem" ... except for a big ol asterisk that actually you can still intentionally guarantee a network policy bypass on deletion, which we have to hope people notice and respect?

It's not worse in that the accidental case is less likely, but we should be careful to not advertise it as a security feature then. And we _are_ when we mention a "vulnerability" to address in the KEP motivation ...

BenTheElder replied:
Can't you already hang deletion on misbehaving finalizers?

Oh no: https://stackoverflow.com/a/59667608

thockin replied:
You CAN hang deletion but everything else will be deleted while the hanging resource hangs.  Slightly (maybe not importantly so) different than this proposal.

danwinship replied:
If our _only_ threat model is "running code in a pod does something bad (or has something bad done to it) after a policy is deleted", then we don't need to make sure the Pod is _deleted_ before the policies are deleted, we just need to make sure it's not `Running`

So the process becomes:
- delete Pods
- wait for Pods to _either_ be completely deleted _or_ reach phase `Succeeded`/`Failed`
- delete rest of namespace

If you put a finalizer on a pod, it prevents it from being fully deleted, but (by design) does not prevent its containers from being stopped, so there will not be any deadlock.

thockin replied:
I'm not sure how to generalize that,  We don't have a mechanism across resources to indicate "finished" except deleted.

danwinship replied:
But do we need to generalize it? Nobody has actually given a concrete example of why we would. (The fact that you can't create new resources in a namespace once you start deleting it is already doing a lot of security work for us.)

thockin replied:
As are you saying to.special-case pods to check for delete OR a terminal state?

Or maybe built-in types can offer a method of self-checking which only Pods use?

BenTheElder replied:
> But do we need to generalize it? Nobody has actually given a concrete example of why we would. (The fact that you can't create new resources in a namespace once you start deleting it is already doing a lot of security work for us.)

I was going to suggest kubevirt type projects with networkpolicy applied to their workload types, but in that particular case kubevirt is running the VM underneath a pod and networkpolicy is targeted via the pod anyhow.

... Also, at some point, trying to defend against every way you can possibly extend things with custom workload types is unreasonable.

Does anyone have a concrete counterpoint? We're already potentially implementing this only for core types anyhow, why not special case pod?

> As are you saying to.special-case pods to check for delete OR a terminal state?
>
> Or maybe built-in types can offer a method of self-checking which only Pods use?

That seems like two ways of saying the same thing while going pretty far into the implementation weeds? Either way "yes" we're specially handling the check for built-in type(s) and _could_ extend it to another type later (but ... it's unclear why we would).

brakthehack replied:
> Well, playing devil's advocate ... It's worse in that people may have a false sense of security that "we fixed the problem" ... except for a big ol asterisk that actually you can still intentionally guarantee a network policy bypass on deletion, which we have to hope people notice and respect?

> It's not worse in that the accidental case is less likely, but we should be careful to not advertise it as a security feature then. And we are when we mention a "vulnerability" to address in the KEP motivation ...

Completely agree. I'd also like to generally ask what is the point of the feature if not for security or consistency? Is preference alone truly valuable to anyone without any strict guarantees?

Special-casing deletion for pods or other resources might be confusing to users. It makes the feature less predictable especially into the future if the progression can be keyed off of any arbitrary resource value. With respect to security, that means the risk for misuse or abuse may increase over time due to expanding complexity.

In addition to good messaging, perhaps allowing users to override deletion order for individual resources could be another option?

BenTheElder replied:
> Special-casing deletion for pods or other resources might be confusing to users. It makes the feature less predictable especially into the future if the progression can be keyed off of any arbitrary resource value. With respect to security, that means the risk for misuse or abuse may increase over time due to expanding complexity.

Workloads get stopped prior to policy deletion is sort of the main intent though? How do you expect this will be abused?

> In addition to good messaging, perhaps allowing users to override deletion order for individual resources could be another option?

Not allowing configuration is intended to prevent expanding complexity for this feature ... I agree it could be useful, but it also seems more ripe for confusion and misuse than just saying "we'll make sure pods stop running before we deleted policies" ?

brakthehack replied:
> Workloads get stopped prior to policy deletion is sort of the main intent though? How do you expect this will be abused?

Not abused per se, but misunderstood. Imagine in the future we have 20 resources that exist in this hierarchy and one or another all have different keys that determine deletion. Even if you have 20 resources and 3-4 of them have different keys. It might be hard for users to remember which conditions allow for deletion.

> Not allowing configuration is intended to prevent expanding complexity for this feature ... I agree it could be useful, but it also seems more ripe for confusion and misuse than just saying "we'll make sure pods stop running before we deleted policies"?

Good points. I guess I assume at some point someone is going to want to treat one or more resources as snowflakes since the deletion order is not set by the user. I don't have a use case in mind at the moment though, so maybe it's best to avoid the premature optimization until a concrete use case arises.

danwinship replied:
> are you saying to special-case pods to check for delete OR a terminal state?

I'm saying that would be the _only_ case. No priorities, no bands, no generic mechanism, no extensibility. When you delete a Namespace, instead of doing "delete all the resources in the namespace (in an undefined order)" it would do:

1. Delete all pods in the namespace (in an undefined order).
2. Wait for all the pods to be stopped or deleted.
3. Delete all the other resources in the namespace (in an undefined order).

That's it. That's the KEP.

We don't need a fully-generic solution just to solve the Pods-vs-NetworkPolicies problem, and [we can't be confident that the proposed generic solution is actually generic enough to solve any problems we don't currently know of](https://github.com/kubernetes/enhancements/pull/5095#discussion_r1936061055).

aojea replied:
I couldn't find a gap to Dan's proposal 🥇

thockin replied:
@cici37 This is a much more limited scope, what do you think?  I am worried that we are solving JUST ONE specific problem, but Dan's right that we don't actuyally know if the more complex solution would solve anything else, because we don't know anything else's problem statements.

Your thoughts?

cici37 replied:
I love @danwinship's proposal. The more order we introduced, the more conflicts it would bring(especially with all the different circumstance of the policy usage and issues it may bring). I'll update the KEP accordingly.

aojea replied:
in spanish the idiom for this is is "muerto el perro se acabó la rabia" , that comes to say that if you remove the origin of the problem you don't have to worry about the effects (something like that)

jpbetz commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+
+#### Story 2 - having ownerReference conflicts with deletion order
+
+When deciding the deletion priority for resources, it should take ownerReference into consideration. 
```
> should take ownerReference into consideration

I believe we're going to be the right be behavior here, but let's spell it out in a bit more detail to make it super clear to readers: If resource X has a ownerReference to Y, and X is higher priority than Y, what is the deletion order between X and Y when the namespace is deleted?  Also, when Y is higher priority than X, what happens?

cici37 replied:
Rephrased and moved it to a separate section.

danwinship commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+Option 2: have the deletion priority bands defined instead of using numbers.
+
+To begin with, the deletion order bands would be introduced:
+- Workload Controllers
```
Does it really matter if you delete workload controllers before workloads? I assume the intent is to stop them from recreating Pods, but once the Namespace has a DeletionTimestamp you can't create new resources in it, right?

thockin replied:
It probably doesn't REALLY matter, but it seems reasonable to have a band before workloads, which is before default.  That said, it could be YAGNI

cici37 replied:
Maybe we could begin with something simply and add order band later only if it's absolutely necessary? I am open with either way

danwinship commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+When deciding the deletion priority for resources, it should take ownerReference into consideration. 
+E.g. the deployment VS pod. However, it should not matter much in terms of namespace deletion. 
+Namespace deletion specifically uses `metav1.DeletePropagationBackground` and all resources would be deleted and the ownerReference 
+dependencies would be handled by the garbage collection.
```
the two "Story 2"s aren't user stories. They're more like "Notes/Constraints/Caveats".

danwinship commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+The existing random deletion process for resources in a namespace poses significant challenges, 
+particularly in environments with strict security requirements. One critical issue is the potential for 
+a Pod to remain active after its associated NetworkPolicy has been deleted, leaving it exposed to unrestricted network access. 
+This creates a security vulnerability during the cleanup process.
```
Though it is considered an incorrect use of NetworkPolicy, there are clusters where the cluster-admins reserve NetworkPolicy for themselves, and use it to enforce policy on users (particularly egress policy). In that case, deleting the NetworkPolicies before the Pods may allow pods to get unrestricted egress access during deletion.

thockin replied:
Let's discuss in https://github.com/kubernetes/enhancements/pull/5095#discussion_r1935990201 ?

danwinship commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+When a lack of progress detected(maybe caused by the dependency cycle described above), the options would be:
+- Fall back to the previous behavior.
+  - Pros: The deletion would not be blocked; no breaking changes;
+  - Cons: Security concern remains unaddressed or could be bypassed
```
In the normal NetworkPolicy case, the user would be bypassing their own security, so... whatever.

But in the "cluster-admin enforcing egress NetworkPolicy on users" case (or any other case involving an admin-created resource vs a user-created resource), this completely breaks the feature.

cici37 replied:
Is this concern related with custom resource which user might create their own policy/workload? We could have the same deletion order applied to CRD and user could configure which deletion group their CR should belong to.

aojea replied:
> But in the "cluster-admin enforcing egress NetworkPolicy on users" case (or any other case involving an admin-created resource vs a user-created resource), this completely breaks the feature.

@danwinship can you elaborate I'm missing the point, if the pod is deleted before the network policy it means that network policies will be applied for the whole live of the pod

danwinship replied:
@aojea I'm saying, if we obey finalizers on Pods, then the plan is basically broken, because the user can always just add a finalizer to subvert the deletion ordering.

thockin replied:
Right.  We were looking at this as an outside attacker hitting the window between NetPol being deleted and Pod being deleted, but Dan points out that the pod owner could use this to hit that window for otherwise disallowed egress!

e.g. admin put a NetPol which says no egress to internet.  The pod gets compromised (or inside-job).  They add an unresolvable finalizer.  When the NS is deleted, the NetPol will eventually go away and the compromised pod can exfiltrate their corporate secrets.

Ouch.  So is that worse than having unresolvable deadlocks?

Perhaps we go back to an earlier comment - we have a timeout for a few releases but then convert to hard-stop?

thockin replied:
xref to https://github.com/kubernetes/enhancements/pull/5095/files#r1935990201 wherein the discussion of cycle-breaking is going on.

aojea replied:
> he pod gets compromised (or inside-job). They add an unresolvable finalizer.

that looks like an rbac problem not a deletion namespace order problem, I don't think we need to build and totally bullet proof system, just put the primitives in place in order to do so ... why a cluster admin will allow people to add finalizer to pods, that means that there has to be some controller removing those ... I think the point is right, but the solution is in another layer

BenTheElder replied:
> that looks like an rbac problem not a deletion namespace order problem, I don't think we need to build a totally bullet proof system, just put the primitives in place in order to do so ... why a cluster admin will allow people to add finalizer to pods, that means that there has to be some controller removing those ... I think the point is right, but the solution is in another layer

I think if we don't, we have to be careful about communicating then. 
If you say "this vulnerability was fixed by this KEP", I expect it to be _fixed_, not partially mitigated with a pretty gaping hole in it.

A cluster admin might reasonably expect that networkpolicies + some sort of other admission controls (prohibit privileged / host net?) would protect them from bad behavior of a controller able to create pods, and a controller creating pods would pretty reasonably want permission to add a finalizer?

BenTheElder replied:
This thread has a better discussion for that problem though: https://github.com/kubernetes/enhancements/pull/5095#discussion_r1935990201

danwinship commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+When a lack of progress detected(maybe caused by the dependency cycle described above), the options would be:
+- Fall back to the previous behavior.
+  - Pros: The deletion would not be blocked; no breaking changes;
+  - Cons: Security concern remains unaddressed or could be bypassed
```
In the case of "Pod outlives policy that restricts the Pod", we don't actually need the Pod to be _deleted_ before we delete the policy, we only need the Pod to be _stopped_...

cici37 replied:
In current namespace deletion process, the resources deletion process would not differentiate `deleting the resource` and `stop the resource`..

thockin commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+- Workload Controllers
+- Workloads
+- Default
+- Policies
```
We can't have infinite bands, and I think we want to discourage trying to get into super fine-grained global order.

We could have polices1...N or have N integer numbers within each band, but I am not convinced it's a good way to spend the complexity budget, in the absence of real use-cases.  If bands map to numbers in the implementation, we can certainly leave room for more.

danwinship replied:
> in the absence of real use-cases

I'm totally willing to believe that this doesn't affect any real-world use cases, but the KEP currently seems to imply that we are worrying about cases we don't know about yet.

Either we only solve the problem of "Pods that are still running when objects that restrict them are deleted", or else we have to figure out specifically what other sorts of problems we are solving. We can't come up with a generic solution that will correctly protect against entirely unknown threats.

thockin replied:
If we allow to break cycles, then I think at least 3 are warranted:

1. pods
2. default
3. policies

This way, a cycle-break on pods moves ahead to default, which might resolve the cycle.  Moving into policies is the last resort.

If decide not to do cycle breaks, then 2 is probably OK, but I think 3 still gives additional warm-and-fuzzy that things which identify as policies come after things which are uknown.

cici37 replied:
Another question is regarding with use case for ValidatingAdmissionPolicy. Say the deletion order of VAP is in `policies`, but VAP is using other resources(e.g. configmap or CRD) for its parameterization which the deletion order is set to `default' and be deleted before, in this case, VAP will be invaluable after the param resource being deleted even through the VAP resource hasn't been deleted yet. Do we have security concern on this case?

thockin replied:
xref https://github.com/kubernetes/enhancements/pull/5095/files#r1935990201

felipe88alves replied:
More than a security concern, what I've found is that an unordered deletion of the parameterization resource can lead the namespace deletion to hang indefinitely.

The namespace hangs if the ValidatingAdmissionPolicy is targeting resources with complex termination flows (such as a PVC and perhaps a LoadBalancer Service) and the associated ValidatingAdmissionPolicyBinding has `parameterNotFoundAction` set to `Deny`.
A reproducible example can be found here: https://github.com/kubernetes/kubernetes/issues/129883

Should we add this UC to the KEP as well?

Note:
Perhaps my comment called for a new thread, please let me know if that's the case

cici37 commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+
+Option 2: have the deletion priority bands defined instead of using numbers.
+
+To begin with, the deletion order bands would be introduced:
```
We could begin with something simpler and add more groups if needed later. I guess both could leave space for future possible deletion order insertion. :)

cici37 commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+
+#### Story 2 - having finalizer conflicts with deletion order
+
+E.g. the NetworkPolicy deletion priority would be set lower than the pod and we could expect the Pod be deleted 
```
Updated with rephrase.

aojea commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
````diff
@@ -0,0 +1,832 @@
+```
+
+After the deletion priority set, the namespace deletion process will honor the priority and delete the resources in sequence. 
+The resources with higher priority should always be deleted before the resources with lower priority.
````
@cici37 seems worth clarifying, since I think that the term "deleted" across the document refers to "absence of object" that means "the thing that was representing the object in the cluster no longer exist" aka "pod deleted and finalized"

aojea commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,832 @@
+specified against deletion priority.
+
+When a lack of progress detected(maybe caused by the dependency cycle described above), the options would be:
+- Fall back to the previous behavior.
```
can we smart and only fall back if finalizers are detected on the objects?

thockin replied:
https://github.com/kubernetes/enhancements/pull/5095/files#r1935990201

aojea commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
```diff
@@ -0,0 +1,834 @@
+In this case, the `NamespaceDeletionOrder` will be associated with the resources naturally and the resource which does not have an opinion will default to a value.
+
+- Pros:
+  - Having it configurable
```
I think of this as can also be a  Con, since exposing the capacity to establish random deletion priorities between different clusters will make the namespace deletion different per cluster

thockin replied:
I don't think anyone wants it to be user-configurable.  The question is whether the type defines it's own classification, or whether we define that external to the type.  I think it is intrinsic to the type, but am not going to die on this hill today.  I feel like types SHOULD BE tightly sealed plugins, and any time we store information about those types outside of the type itself, it's a smell.

BenTheElder replied:
> I don't think anyone wants it to be user-configurable. The question is whether the type defines it's own classification, or whether we define that external to the type. I think it is intrinsic to the type, but am not going to die on this hill today. I feel like types SHOULD BE tightly sealed plugins, and any time we store information about those types outside of the type itself, it's a smell.

When you say type, only core types, not CRDs?

thockin replied:
For now core, but when CRDs get this, they would be the same.  CRDs which implement policies could denote that.

BenTheElder replied:
That's at least one bit of user config? :-)

Point taken though, we should probably at least spell out how we're leaving room to add that later without resolving details such as "what is the CRD opt-in field" yet?

aojea commented on keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md:
````diff
@@ -0,0 +1,834 @@
+- Cons:
+  - Hard to oversee the overall prioritization across resources
+
+2. Maintain a hard-coded map of DeletionOrder per resources for the resources which have a deletion order preference. Any resources which have no preference opinion would be in the Default category.
+```
+var resourceDeletionOrderPriorities = map[ResourceType]DeletionOrderPriority{
+	"/pods": DeletionOrdeWorkload,
+	"apps.k8s.io/deployments":  DeletionOrdeWorkloadController.
+	"networking.k8s.io/networkpolicies”: DeletionOrdePolicy,
+	……
+})
````
we already have this model in the apiserver aggregator https://github.com/kubernetes/kubernetes/blob/2e64c722492c9bbb54b177258e7b659717f7e3da/cmd/kube-apiserver/app/aggregator.go

```
/ The proper way to resolve this letting the aggregator know the desired group and version-within-group order of the underlying servers
// is to refactor the genericapiserver.DelegationTarget to include a list of priorities based on which APIs were installed.
// This requires the APIGroupInfo struct to evolve and include the concept of priorities and to avoid mistakes, the core storage map there needs to be updated.
// That ripples out every bit as far as you'd expect, so for 1.7 we'll include the list here instead of being built up during storage.
var apiVersionPriorities = merge(controlplaneapiserver.DefaultGenericAPIServicePriorities(), map[schema.GroupVersion]controlplaneapiserver.APIServicePriority{
	{Group: "", Version: "v1"}: {Group: 18000, Version: 1},
	// to my knowledge, nothing below here collides
	{Group: "apps", Version: "v1"}:                    {Group: 17800, Version: 15},
	{Group: "autoscaling", Version: "v1"}:             {Group: 17500, Version: 15},
	{Group: "autoscaling", Version: "v2"}:             {Group: 17500, Version: 30},
	{Group: "autoscaling", Version: "v2beta1"}:        {Group: 17500, Version: 9},
	{Group: "autoscaling", Version: "v2beta2"}:        {Group: 17500, Version: 1},
	{Group: "batch", Version: "v1"}:                   {Group: 17400, Version: 15},
	{Group: "batch", Version: "v1beta1"}:              {Group: 17400, Version: 9},
	{Group: "batch", Version: "v2alpha1"}:             {Group: 17400, Version: 9},
	{Group: "networking.k8s.io", Version: "v1"}:       {Group: 17200, Version: 15},
	{Group: "networking.k8s.io", Version: "v1beta1"}:  {Group: 17200, Version: 9},
	{Group: "networking.k8s.io", Version: "v1alpha1"}: {Group: 17200, Version: 1},
	{Group: "policy", Version: "v1"}:                  {Group: 17100, Version: 15},
	{Group: "policy", Version: "v1beta1"}:             {Group: 17100, Version: 9},
	{Group: "storage.k8s.io", Version: "v1"}:          {Group: 16800, Version: 15},
	{Group: "storage.k8s.io", Version: "v1beta1"}:     {Group: 16800, Version: 9},
	{Group: "storage.k8s.io", Version: "v1alpha1"}:    {Group: 16800, Version: 1},
	{Group: "scheduling.k8s.io", Version: "v1"}:       {Group: 16600, Version: 15},
	{Group: "node.k8s.io", Version: "v1"}:             {Group: 16300, Version: 15},
	{Group: "node.k8s.io", Version: "v1alpha1"}:       {Group: 16300, Version: 1},
	{Group: "node.k8s.io", Version: "v1beta1"}:        {Group: 16300, Version: 9},
	{Group: "resource.k8s.io", Version: "v1beta1"}:    {Group: 16200, Version: 9},
	{Group: "resource.k8s.io", Version: "v1alpha3"}:   {Group: 16200, Version: 1},
	// Append a new group to the end of the list if unsure.
	// You can use min(existing group)-100 as the initial value for a group.
	// Version can be set to 9 (to have space around) for a new group.
})
```

I think this is simple and easier to maintain in the long term, as the motivation for this is NetworkPolicies and Pods, do we envision to have more use cases to establish priorities on namespace deletion?

BenTheElder replied:
> I think this is simple and easier to maintain in the long term, as the motivation for this is NetworkPolicies and Pods, do we envision to have more use cases to establish priorities on namespace deletion?

I think this is a key outstanding question, which is ongoing in a thread above.

cici37 replied:
I guess this is discussed in above comment thread but wanna point out the main benefit we could get from configuring the deletion order through APIResource is the consistency between native type and CRD.

aojea commented:
lol, we have a loop on the comments links, so you get redirected to one thread from another thread that brings you back to the original thread 🤣

danwinship commented:
shh! it's a trap for AI scrapers!

cici37 commented:
Summary on current discussions:

1. Thanks for @danwinship 's proposal, I have updated the KEP to reflect [his comment](https://github.com/kubernetes/enhancements/pull/5095#discussion_r1940385596)
  ```
   - Delete all pods in the namespace (in an undefined order).
   - Wait for all the pods to be stopped or deleted.
   - Delete all the other resources in the namespace (in an undefined order).
  ```
2. I have added a section for security concern regarding with policy with parameter resources.([section](https://github.com/kubernetes/enhancements/blob/5f6e3879593594481f3f1f9580c2a128d25f0276/keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md#story-3---having-policy-set-up-with-parameter-resources), [related issue](https://github.com/kubernetes/kubernetes/issues/129883)). However, due to the changes in 1, this will be out of the scope of current KEP.
3. In this case, we would like to have the deletion process described in 1 being strict enforced with a feature gate guarded.

Thank you so much for the input!

felipe88alves commented:
Apologies if this has already been discussed, but isn't the idea of deleting the ValidatingAdmissionPolicy and ValidatingAdmissionPolicyBinding resources last simple enough to add here?
Seems like this would solve problem the parameter resource problem.

```
 - Delete all pods in the namespace (in an undefined order).
 - Wait for all the pods to be stopped or deleted.
 - Delete all the other resources in the namespace except ValidatingAdmissionPolicy and ValidatingAdmissionPolicyBinding (in an undefined order).
 - Wait for all the resources to be stopped or deleted.
 - Delete all the ValidatingAdmissionPolicy and ValidatingAdmissionPolicyBinding resources.
 ```

cici37 commented:
> Apologies if this has already been discussed, but isn't the idea of deleting the ValidatingAdmissionPolicy and ValidatingAdmissionPolicyBinding resources last simple enough to add here? 

It will not address the issues of:
1. Security concern regarding with VAP being invalid after parameter resources being deleted
2. VAP using parameter resource with `.spec.paramRef.parameterNotFoundAction: Deny`. VAP could even use pod for parameter resource(Even through we would not expect this much in real use case but there is no technical blocker to prevent user from doing so). It would cause blocker for the followup deletion.
3. VAP is cluster scoped and is not limited on guard resources in current namespace. 
4. The similar concerns exist also with Webhooks using parameter resources.

The current proposal may not be perfect for all scenarios but at least would not make things worse and would address the security concern of workload in a reasonable way.

thockin reviewed (commented):
/lgtm
/approve

cici37 commented:
/assign @jpbetz 
for PRR and KEP approval. Thank you!

jpbetz commented:
/approve
For PRR
